src/api/github/push.ts

```diff
@@ -0,0 +1,198 @@
+import fs from 'fs/promises';
+
+import { graphql } from '@octokit/graphql';
+import type {
+  CreateCommitOnBranchInput,
+  FileAddition,
+  FileDeletion,
+} from '@octokit/graphql-schema';
+
+import * as Git from '../git';
+import type { ChangedFile } from '../git/getChangedFiles';
+
+import { apiTokenFromEnvironment } from './environment';
+
+interface CreateCommitResult {
+  createCommitOnBranch: {
+    commit: {
+      id: string;
+    };
+  };
+}
+
+interface CommitAndPushAllChangesParams {
+  dir: string;
+  /**
+   * The branch name
+   */
+  branch: string;
+  /**
+   * The headline of the commit message
+   */
+  messageHeadline: string;
+  /**
+   * The body of the commit message
+   */
+  messageBody?: string;
+  /**
+   * Updates the local git working directory to reflect the new remote state
+   */
+  updateLocal?: boolean;
+}
+
+export const commitAndPushAllChanges = async ({
+  dir,
+  branch,
+  messageHeadline,
+  messageBody,
+  updateLocal,
+}: CommitAndPushAllChangesParams): Promise<string> => {
+  const changedFiles = await Git.getChangedFiles({ dir });
+  const fileChanges = await mapChangedFilesToFileChanges(changedFiles);
+
+  return await commitAndPush({
+    dir,
+    branch,
+    messageHeadline,
+    messageBody,
+    fileChanges,
+    updateLocal,
+  });
+};
+
+interface FileChanges {
+  additions: FileAddition[];
+  deletions: FileDeletion[];
+}
+
+export const mapChangedFilesToFileChanges = async (
+  changedFiles: ChangedFile[],
+): Promise<FileChanges> => {
+  const { added, deleted } = changedFiles.reduce<{
+    added: string[];
+    deleted: string[];
+  }>(
+    (files, changedFile) => {
+      const filePath = changedFile.path;
+      if (changedFile.state === 'deleted') {
+        files.deleted.push(filePath);
+      } else {
+        files.added.push(filePath);
+      }
+
+      return files;
+    },
+    { added: [], deleted: [] },
+  );
+
+  const additions: FileAddition[] = await Promise.all(
+    added.map(async (filePath) => ({
+      path: filePath,
+      contents: await fs.readFile(filePath, {
+        encoding: 'base64',
+      }),
```

Comment:
Is there a request size limit that we need to be aware of here? In that scenario, does GitHub provide an option to upload files separately? (I'm hoping we can avoid this complexity, but just thinking out loud.)

Reply:
None that they publicly say. Does not look like there's an option to upload files separately unless you want to use their REST API which requires about 3 different API calls to make a commit :\